### PR TITLE
Draft: Miri/CTFE engine: Unions are not scalar

### LINKED ISSUE #69488: const-eval: load of partially initialized scalar produces entirely uninitialized result
The constant evaluation of a move of a union differs from the semantics of non-constant evaluation. The union instance is not regarded as a bag-of-bits as aluded to in the reference and also [this `unsafe-code-guidelines` issue](https://github.com/rust-lang/unsafe-code-guidelines/issues/73#issue-397775312). This occurs only if the unions has a single scalar *scalar* field and otherwise ZSTs, which is pretty specific but occurs for `MaybeUninit` if the type parameter is a scalar. If the field is of other types such as arrays (`[scalar; N]`) *or* if the other field is not a ZST then the semantics seem fine.

I tried this code:

```rust
#![feature(const_if_match)] // Only related to make_1u8_bag, all of them set by Miri
#![feature(const_fn)]
#![feature(const_mut_refs)]
#![feature(const_panic)]
#![feature(const_raw_ptr_deref)]

// Or, equivalently: `MaybeUninit`.
pub union BagOfBits<T: Copy> {
    uninit: (),
    _storage: T,
}

pub const fn make_1u8_bag<T: Copy>() -> BagOfBits<T> {
    assert!(core::mem::size_of::<T>() >= 1);
    let mut bag = BagOfBits { uninit: () };
    unsafe { (&mut bag as *mut _ as *mut u8).write(1); };
    bag
}

pub fn check_bag<T: Copy>(bag: &BagOfBits<T>) {
    let val = unsafe { (bag as *const _ as *const u8).read() };
    assert_eq!(val, 1);
}

fn main() {
    check_bag(&make_1u8_bag::<[usize; 1]>()); // Fine
    check_bag(&make_1u8_bag::<usize>()); // Fine

    const CONST_ARRAY_BAG: BagOfBits<[usize; 1]> = make_1u8_bag();
    check_bag(&CONST_ARRAY_BAG); // Fine.
    const CONST_USIZE_BAG: BagOfBits<usize> = make_1u8_bag();
    check_bag(&CONST_USIZE_BAG); // Panics
}
```

[Playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=deecbf5c59706a37f7db16ea5f3ed30b)

I expected to see this happen: All assertions succeed.

Instead, this happened: The check where a union of type `BagOfBits<usize>` is *moved during constant evaluation*, panics. When putting everything through Miri it panics with the message of trying to read an uninitialized byte. The check where the union of type `BagOfBits<[usize; 1]>` is used instead runs fine in both cases.

```text
error: Miri evaluation error: attempted to read undefined bytes
  --> src/main.rs:21:5
   |
21 |     assert_eq!(val, 1);
   |     ^^^^^^^^^^^^^^^^^^^ attempted to read undefined bytes
```

As far as I am aware, when constant evaluation reads *scalars* then it propagates uninitialized bytes. If that were to happen then it would explain the observed behaviour but it seems weird why the union copy would use a scalar read.

cc @RalfJung

### PR DESCRIPTION
**Pull request by RalfJung**
To fix https://github.com/rust-lang/rust/issues/69488 and by extension https://github.com/rust-lang/rust/issues/94371, we should stop treating types like `MaybeUninit<usize>` as something that the `Scalar` type in the interpreter engine can represent. So we add a new test `type_is_scalar` which checks if a type can be represented as a `Scalar`.

This is a draft since it still ICEs in Miri, but I need help figuring out the best way to avoid that.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -246,6 +246,57 @@ impl<'tcx, Tag: Provenance> ImmTy<'tcx, Tag> {
...
+                // Check all fields of all variants, to make sure there is no union hiding here.
+                adt_def.all_fields().all(|field| type_is_scalar(tcx, field.ty(tcx, substs)))
+            }
+        }
+        ty::Tuple(..) => {
+            // If all fields are scalar, we are good.
+            ty.tuple_fields().iter().all(|field| type_is_scalar(tcx, field))
+        }
+        // FIXME: can these ever have Scalar ABI?
+        ty::Closure(..) | ty::Generator(..) => false,
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:282]: This line is the problem: some closures (and likely generators) *can* end up with a `Scalar` layout, and they can even have niches, which then means validity will call `read_scalar` and cause an ICE when we do not treat it as a scalar.

To fix this, we need a way to iterate over the types of all the fields of (all the variants of) a closure/generator. Any suggestion for how to do that? :D I browsed the code for a bit and found nothing.

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/operand.rs:282]: For closures you'll want https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.ClosureSubsts.html#method.upvar_tys which you get by copying the substs into a new instance of ClosureSubsts. The same helper exists for generators. I guess we should make this less cumbersome to do

**Review by oli-obk — COMMENTED** [review]
I'm not too happy with this approach. Maybe we should add a flag to scalar layout information stating whether it has a union inside? Maybe other backends are interested in whether a scalar can be undef, too?

**Comment by RalfJung** [comment]
> I'm not too happy with this approach. Maybe we should add a flag to scalar layout information stating whether it has a union inside? Maybe other backends are interested in whether a scalar can be undef, too?

I agree looking at the types is somewhat icky, and ideally `abi::Scalar` would contain the needed information. But that code is quite outside what I know how to handle...

FWIW, https://github.com/rust-lang/rust/pull/93670 actually has some overlap with this I think. There, a new attribute was added to `ArgAttributes` -- but that only helps for function calls.

And indeed I think LLVM might some day be interested in this distinction -- `i64` cannot be partially poison, so if Rust code can ever generate `poison` values (I think currently it cannot), then `MaybeUninit<u64>` cannot be compiled to just LLVM `i64` any more, it would have to be something involving the "byte type" that was proposed to be added to LLVM.

**Comment by RalfJung** [comment]
> For closures you'll want https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.ClosureSubsts.html#method.upvar_tys which you get by copying the substs into a new instance of ClosureSubsts. The same helper exists for generators. I guess we should make this less cumbersome to do

(I cannot reply directly to that comment, WTF GitHub?)

I was trying to figure out how to construct a ClosureSubst, but `ClosureSubst::new` did not help. Should I just construct a struct directly? But then I would be entirely ignoring the `DefId` in `ty::Closure`, which seems rather suspicious? At that point I decided this is probably wrong and gave up.

**Comment by oli-obk** [comment]
I can check on Tuesday, but my approach is basically "check other use sites, do the same thing and document the findings and/or add helpers"

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 2]
```diff
@@ -246,6 +246,57 @@ impl<'tcx, Tag: Provenance> ImmTy<'tcx, Tag> {
...
+        // Types that don't have scalar layout to begin with.
+        ty::Array(..) | ty::Slice(..) | ty::Str | ty::Dynamic(..) | ty::Foreign(..) => {
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs]: At least currently, arrays [always seem to have `Aggregate` ABI](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6092dffeefea311da6219a2d4759c1c2).

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_const_eval/src/interpret/operand.rs:282]: What about renaming this to `type_can_be_scalar` and returning true in all cases except unions?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs:282]: Then we'd still have the bug for (e.g.) closures that capture unions in a way that they still get `abi::Scalar` layout.

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/operand.rs:282]: I am experimenting with changing the layout to support undef scalars. If perf is happy with it, we can just ask the layout whether the scalars are undef and not do an immediate representation then.

**Comment by RalfJung** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/94527.

**Pull request closed without merging**